### PR TITLE
Refactor CArtPlace to create CCommanderArtPlace + extend commander artifact support

### PR DESCRIPTION
**Pull request by dydzio0614**
Second commit probably makes CCommanderArtPlace more ugly and responsible for more things than it should, but it allows to give commander artifacts back to hero via left click on artifact.

### CODE REVIEW

File: client/widgets/CArtifactHolder.cpp [hunk 1]
```diff
@@ -945,3 +943,102 @@ bool CArtifactsOfHero::SCommonPart::Artpos::valid()
...
+	ArtifactPosition freeSlot = ourArt->firstBackpackSlot(commanderOwner);
+
+	ArtifactLocation src(commanderOwner->commander.get(), artifactPos);
+	ArtifactLocation dst(commanderOwner, freeSlot);
+
+	if (ourArt->canBePutAt(dst, true))
+	{
+		LOCPLINT->cb->swapArtifacts(src, dst);
+		setArtifact(nullptr);
+		parent->redraw();
```
Comment (dydzio0614) [inline client/widgets/CArtifactHolder.cpp]: This line is probably a code smell, but I did not find better solution with current design.

File: client/widgets/CArtifactHolder.cpp [hunk 2]
```diff
@@ -945,3 +943,102 @@ bool CArtifactsOfHero::SCommonPart::Artpos::valid()
...
+	}
+}
+
+void CCommanderArtPlace::setArtifact(const CArtifactInstance * art)
```
Comment (dydzio0614) [inline client/widgets/CArtifactHolder.cpp]: This method is similar to CHeroArtPlace::setArtifact - code duplication?

File: client/widgets/CArtifactHolder.h [hunk 3]
```diff
@@ -41,13 +41,44 @@ class CWindowWithArtifacts : public CArtifactHolder
...
+	void clickRight(tribool down, bool previousState) override;
+
+	virtual void setArtifact(const CArtifactInstance *art)=0;
+};
+
+class CCommanderArtPlace : public CArtPlace
+{
+protected:
+	const CGHeroInstance * commanderOwner;
+	ArtifactPosition commanderSlotID;
```
Comment (dydzio0614) [inline client/widgets/CArtifactHolder.h]: I decided to separate this from CHeroArtPlace::slotID, as the latter is managed magically by CArtifactsOfHero, and I did not want to set it up in CArtPlace constructor (which is now a base class)

File: client/widgets/CArtifactHolder.cpp [hunk 4]
```diff
@@ -945,3 +943,102 @@ bool CArtifactsOfHero::SCommonPart::Artpos::valid()
...
+{
+	LRClickableAreaWTextComp::clickLeft(down, previousState);
+}
+
+void CArtPlace::clickRight(tribool down, bool previousState)
+{
+	LRClickableAreaWTextComp::clickRight(down, previousState);
+}
+
+CCommanderArtPlace::CCommanderArtPlace(Point position, const CGHeroInstance * commanderOwner, ArtifactPosition artSlot, const CArtifactInstance * Art) : CArtPlace(position, Art), commanderOwner(commanderOwner), commanderSlotID(artSlot.num)
```
Comment (dydzio0614) [inline client/widgets/CArtifactHolder.cpp]: Yeah, this class constructor takes may arguments, but I guess it is ok, considering it does everything without relying on CArtifactsOfHero magic

File: client/widgets/CArtifactHolder.cpp [hunk 5]
```diff
@@ -945,3 +943,102 @@ bool CArtifactsOfHero::SCommonPart::Artpos::valid()
...
+{
+	createImage();
+	setArtifact(Art);
+}
+
+void CCommanderArtPlace::clickLeft(tribool down, bool previousState)
+{
+	if (down && ourArt && text.size())
+		LOCPLINT->showYesNoDialog("Do you want to give this artifact back to hero?", [this] { returnArtToHeroCallback(); }, [] {});
```
Comment (alexvins) [inline client/widgets/CArtifactHolder.cpp]: Missing localization. Use config/traslate.json

**Pull request merged**